### ISSUE TITLE
Leaking peer connections on disconnect

### ISSUE THREAD

**Issue by thetooth** [pions/webrtc#108]
Currently having an issue where a long running process will eventually run out of file descriptors because the UDP socket is left open when the peer is no longer listening but state is stuck as "connected".

As I am building a SFU, we use a async broker to fanout a media stream to multiple peers, when the track offered to the peer stops accepting samples on it's channel, I determine the peer to be gone as blocking here would stop samples for every other peer, instead the sample is dropped and I call Close() on the peer connection to clean it up, this however also blocks and remains blocked until the application is terminated.

The block is occurring here: https://github.com/pions/webrtc/blob/master/internal/network/manager.go#L120
As it tries to acquire a lock, I see this mutex used all over the place, namely in SendRTP. I believe we're getting stuck here trying to send to a peer that is no longer listening, making closing and freeing the socket impossible.

**Comment by thetooth** [comment]
SFU is "working" because we're taking a naive approach, it's not correct. FIR or PLI are not strictly necessary with VP8 as the client can figure out the frame from any starting point.

As for reproducing, I can replicate this using the gstreamer send example simply by modifying the client to refresh on disconnect(which occurs often because we're testing behind a NAT that STUN has no chance of resolving) and letting it run for a few hours/smashing F5.

Where it actually stops would seem to be inside the /x/ipv4 library... maybe we should be calling SetWriteDeadline?

**Comment by thetooth** [comment]
Nice! Good to hear this will be addressed, it was a pretty non-obvious problem, I was convinced I was just using the library wrong for some time.

I have done some quick tests and the patch corrects the problem fully, connection durability is suddenly 100% on LAN which it was not before, I had originally assumed that was due to a broken ICE configuration(STUN cannot resolve our topology, waiting on TURN). Peers that leave normally are being closed within 15 seconds or so, and the edge case where the RTCTrack channel blocked is no longer happening, as I suspected that was just a symptom of the underlying problem.

I'll let you know tomorrow if there are any regressions, both on the version were using (41e7adbb2bbc3e984464286ede649148a699f6a3) and head.
